pkg/service/controller/servingcert/controller.go

```diff
@@ -0,0 +1,232 @@
+func (sc *ServiceServingCertController) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	go sc.serviceController.Run(stopCh)
+	for i := 0; i < workers; i++ {
+		go wait.Until(sc.worker, time.Second, stopCh)
+	}
+	<-stopCh
+	glog.Infof("Shutting down service signing cert controller")
+	sc.queue.ShutDown()
+}
+
+func (sc *ServiceServingCertController) enqueueService(obj interface{}) {
+	key, err := controller.KeyFunc(obj)
+	if err != nil {
+		glog.Errorf("Couldn't get key for object %+v: %v", obj, err)
+		return
+	}
+
+	sc.queue.Add(key)
+}
+
+// worker runs a worker thread that just dequeues items, processes them, and marks them done.
+// It enforces that the syncHandler is never invoked concurrently with the same key.
+func (sc *ServiceServingCertController) worker() {
+	for {
+		func() {
+			key, quit := sc.queue.Get()
+			if quit {
+				return
+			}
+			defer sc.queue.Done(key)
+			err := sc.syncHandler(key.(string))
+			if err != nil {
+				glog.Errorf("Error syncing service: %v", err)
+			}
+		}()
+	}
+}
+
+// syncService will sync the service with the given key.
+// This function is not meant to be invoked concurrently with the same key.
+func (sc *ServiceServingCertController) syncService(key string) error {
+	obj, exists, err := sc.serviceCache.GetByKey(key)
+	if err != nil {
+		glog.Infof("Unable to retrieve service %v from store: %v", key, err)
+		sc.queue.Add(key)
+		return err
+	}
+	if !exists {
+		glog.Infof("Service has been deleted %v", key)
+		return nil
+	}
+
+	// make a copy to avoid mutating cache state
+	t := *obj.(*kapi.Service)
+	service := &t
+
+	if !sc.requiresCertGeneration(service) {
+		return nil
+	}
+
+	dnsName := service.Name + "." + service.Namespace + ".svc"
+	fqDNSName := dnsName + "." + sc.dnsSuffix
+	servingCert, err := sc.ca.MakeServerCert(sets.NewString(dnsName, fqDNSName))
+	if err != nil {
+		return err
+	}
+	certBytes, keyBytes, err := servingCert.GetPEMBytes()
+	if err != nil {
+		return err
+	}
+
+	secret := &kapi.Secret{}
+	secret.Name = service.Annotations[ServingCertSecretAnnotation]
+	secret.Namespace = service.Namespace
+	secret.Type = kapi.SecretTypeTLS
+	secret.Data = map[string][]byte{}
+	secret.Data[kapi.TLSCertKey] = certBytes
+	secret.Data[kapi.TLSPrivateKeyKey] = keyBytes
+	secret.Annotations = map[string]string{}
+	secret.Annotations[ServiceUIDAnnotation] = string(service.UID)
+	secret.Annotations[ServiceNameAnnotation] = service.Name
+	_, err = sc.secretClient.Secrets(service.Namespace).Create(secret)
+	if err != nil && !kapierrors.IsAlreadyExists(err) {
+		// if we have an error creating the secret, then try to update the service with that information.  If it fails,
+		// then we'll just try again later on  re-list or because the service had already been updated and we'll get triggered again.
+		service.Annotations[ServingCertErrorAnnotation] = err.Error()
+		sc.serviceClient.Services(service.Namespace).Update(service)
+		return err
```

Comment:
> Enqueue the service in such a case?

That risks hot-loops on failures.  Perhaps a followup that looses the "cert generation failures are never retried" to allow "X retries before perm failing" could add that sort of logic?